File: 프로그래머스/lv3/42895. N으로 표현/N으로 표현.java
import java.util.*;

class Solution {
    public int solution(int N, int number) {
        List<Set<Integer>> setList = new ArrayList<>();
        
        for (int i=0; i<9; i++) {
            setList.add(new HashSet<Integer>());
        }
        
        setList.get(1).add(N);
        
        for (int i=2; i<9; i++) {
            Set<Integer> currentSet = setList.get(i);
            
            for (int j=1; j<i; j++) {
                Set<Integer> preSet = setList.get(j);
                Set<Integer> postSet = setList.get(i-j);
                
                for (Integer pre : preSet) {
                    for (Integer post : postSet) {
                        currentSet.add(pre + post);
                        currentSet.add(pre * post);
                        currentSet.add(pre - post);
                        if (pre != 0 && post != 0) {
                            currentSet.add(pre / post);
                            currentSet.add(post / pre);
                        }
                    }
                }
            }
            
            currentSet.add(Integer.parseInt(String.valueOf(N).repeat(i)));
        }
        
        for (Set set : setList) {
            if (set.contains(number)) {
                return setList.indexOf(set);
            }
        }
        
        return -1;
    }
}

File: 프로그래머스/lv3/42895. N으로 표현/N으로 표현.kt
class Solution {
    fun solution(N: Int, number: Int): Int {
        val setList: MutableList<MutableSet<Int>> = mutableListOf()
        
        for (i in 0 .. 7) {
            setList.add(mutableSetOf())
        }
        
        for (i in 0 until setList.size) {
            var value = 0
            
            for (j in 0 .. i) {
                value += (Math.pow(10.0, j.toDouble()) * N).toInt()
            }
            
            setList.get(i).add(value)
        }
        
        for (i in 1 .. 7) {
            val currentSet = setList.get(i)
            val tempSet: MutableSet<Int> = mutableSetOf()
            
            for (j in 0 until i) {
                val prevSet = setList.get(j)
                val postSet = setList.get(i-j-1)
                
                for (p1 in prevSet) {
                    for (p2 in postSet) {
                        
                        val n1 = p1 + p2
                        val n2 = p1 * p2
                        val n3 = p1 - p2
                        
                        tempSet.add(n1)
                        tempSet.add(n2)
                        tempSet.add(n3)
                        
                        if (p1 != 0 && p2 != 0) {
                            val n4 = p1/p2
                            val n5 = p2/p1
                            
                            tempSet.add(n4)
                            tempSet.add(n5)
                        }
                    }
                }
            }
            
            currentSet.addAll(tempSet)
        }
        
        for (i in setList.indices) {
            
            if (setList.get(i).contains(number)) {
                return i + 1
            }
        }
        
        return -1
    }
}